src/data/data.cc

```diff
@@ -865,6 +866,87 @@ uint64_t SparsePage::Push(const AdapterBatchT& batch, float missing, int nthread
   return max_columns;
 }
 
+template<>
+uint64_t SparsePage::Push<xgboost::data::DenseAdapterBatch>(
+               const xgboost::data::DenseAdapterBatch& batch,
+               float missing, int nthread, int nthread_original) {
```

Comment:
We can make the changes in general without templates specification.

Reply:
unfortunately there is no possibility to keep obtained speed up for unified version.
The problem is: 
`builder.InitBudget(expected_rows, nthread);` allocates `nthread` times bigger buffer (on datasets with small number of columns like higgs it becomes the hot spots and no performance improvements are observed).

To eliminate described problem in specification optimization was possible by taking into account batch adapter type (threading by rows was implemented), it won't be applicable for `CSCAdapterBatch`.

Reply:
But general case will be also optimized :)

Reply:
If you need to change only one call depending on template type - you can use small separate template function which will call InitBudget in appropriate manner based on Adapter type.

Reply:
unified approach was implemented.